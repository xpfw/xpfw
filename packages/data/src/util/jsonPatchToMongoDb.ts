const toDot = (path: any) => path.replace(/^\//, "").replace(/\//g, ".")

const jsonPatchToMongoDb = (patches: any[]) => {
  const update: any = {}
  patches.map((p) => {
    if (p.op === "replace" || p.op === "add") {
      if (!update.$set) { update.$set = {} }
      update.$set[toDot(p.path)] = p.value
    } else if (p.op === "remove") {
      if (!update.$unset) { update.$unset = {} }
      update.$unset[toDot(p.path)] = 1
    }
  })
  return update
}

export default jsonPatchToMongoDb
