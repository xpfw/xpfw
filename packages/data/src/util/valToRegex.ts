import { cloneDeep } from "lodash"

const valToRegex = (val: any) => {
  if (val == null || val.length === 0) {
    return undefined
  }
  return {
    $regex: `(.*?)${val}(.*?)`,
    $options: "isg"
  }
}

const changeValToRegex = (path: string) => {
  return (value: any) => {
    if (value[path] != null) {
      const newValue = cloneDeep(value)
      newValue[path] = valToRegex(newValue[path])
      return newValue
    }
    return value
  }
}

export default valToRegex
export {
  changeValToRegex
}
