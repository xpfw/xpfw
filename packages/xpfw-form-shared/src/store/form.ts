import {
  globals, IField, IFieldError, IForm, IFormError, iterateFields,
  validateField, validateForm, ValidationRegistry
} from "@xpfw/validate"
import { get, isFunction, isNil, isObject, set } from "lodash"
import { autoSubscribe, AutoSubscribeStore, key, StoreBase } from "resub"
import { IPersistableStore } from "resub-persist"
import FormErrorStore from "./error"

@AutoSubscribeStore
export class FormStore extends StoreBase implements IPersistableStore {
  public name = "formData"
  private formData: any = {}

  public getPropKeys() { return ["formData"] }

  public setValue(valuePath: string, value: any, fullTrigger: boolean = false) {
    set(this.formData, valuePath, value)
    this.trigger(fullTrigger ? undefined : valuePath)
  }

  @autoSubscribe
  public getValue(@key valuePath: string): any {
    return get(this.formData, valuePath)
  }

  public getFormData(form: IForm, prefix: string = "", method: string = "create") {
    const fullData: any = {}
    const formData = this.formData
    iterateFields(form, (field) => {
      let defaultValue = get(field, "validate.defaultValue")
      if (method === globals.Method.Find) {
        defaultValue = undefined
      }
      let value = get(formData,
        `${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, defaultValue)
      let converter: any = get(field, "validate.convert")
      if (!isFunction(converter) && isObject(converter)) {
        converter = converter[method]
      }
      if (isFunction(converter)) {
        value = converter(value)
      }
      if (method !== globals.Method.Find || method === globals.Method.Find && !isNil(value))  {
        set(fullData, field.mapTo, value)
      }
    })
    return fullData
  }

  public copyDocToFormData(doc: any, form: IForm, prefix: string = "") {
    iterateFields(form, (field) => {
        this.setValue(`${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, get(doc, field.mapTo))
    })
    this.trigger()
  }

  public async validateField(field: IField, form: IForm, method: string = "create",
                             prefix: string = "", currentUser?: string) {
    const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`
    try {
      await validateField(get(this.formData, valuePath), field, {
        method,
        field,
        form,
        currentUser,
        registry: ValidationRegistry
      })
      FormErrorStore.setError(valuePath, {ok: true})
    } catch (e) {
      FormErrorStore.setError(valuePath, e)
    }
  }

  public async validateForm(form: IForm, method: string = "create", prefix: string = "", currentUser?: string) {
    const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${form.model}`
    try {
      await validateForm(this.getFormData(form, prefix), form, {
        method,
        form,
        currentUser,
        registry: ValidationRegistry
      })
      FormErrorStore.setError(valuePath, {ok: true})
    } catch (e) {
      FormErrorStore.setError(valuePath, e)
    }
  }

  public resetForm(form: IForm, prefix: string = "")  {
    iterateFields(form, (field) => {
      this.setValue(`${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, undefined)
    })
  }
}

export default new FormStore()
