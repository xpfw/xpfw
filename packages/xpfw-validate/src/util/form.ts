import { isObject } from "lodash"
import { IField, IForm } from "../typeDef"

const getFieldsFromForm: (form: IForm) => IField[] = (form: IForm) => {
  const fields: IField[] = []
  if (isObject(form)) {
    for (const section of form.sections) {
      for (const field of section.fields) {
        fields.push(field)
      }
    }
  }
  return fields
}

export { getFieldsFromForm }
