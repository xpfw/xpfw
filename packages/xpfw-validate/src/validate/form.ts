import { get, indexOf, intersection, isNil, isObject, isString, keys, set } from "lodash"
import { ErrorType, Method } from "../globals"
import getPermittedFields from "../permission/getPermittedFields"
import { IForm, IFormError, IParametersBase } from "../typeDef"
import validateField from "./any"
import validateQueryObject from "./queryObject"
import checkForFieldInSubKeys from "./util/checkForFieldInSubKeys"
import getRequireOwner from "./util/getRequireOwner"

const updateArrayOperators: string[] = [
  "$", "$inc", "addToSet", "$pop", "$pullAll", "$pull", "$pushAll", "$push",
  "$each", "$slice", "$sort", "$position", "$inc", "$mul", "$rename", "$setOnInsert",
  "$set", "$unset", "$min", "$max", "$currentDate", "$limit", "$skip", "$or", "$and"
]

const validateForm: (value: any, form: IForm, params: IParametersBase) => Promise<any> =
async (value, form, params) => {
  if (isObject(params)) {
    if (isNil(params.form)) {
      params.form = form
    }
    switch (params.method) {
      case Method.Find: {
        return validateQueryObject(value, form, params)
        break
      }
      case Method.Update: {
        try {
          const newRes = await checkForFieldInSubKeys(value, form, params, updateArrayOperators)
          return newRes
        } catch (e) {
          if (e !== -1) {
            throw e
          }
        }
        break
      }
    }
  }
  const validatedObject = {}
  const fields = await getPermittedFields(form, params, true)
  if (fields.length === 0) {
      return Promise.resolve(validatedObject)
  }
  const formError: IFormError = {
    model: form.model,
    errors: []
  }

  for (const field of fields) {
    try {
      const newVal = await validateField(get(value, field.mapTo), field, params)
      if (!isNil(newVal)) {
        set(validatedObject, field.mapTo, newVal)
      }
    } catch (e) {
      formError.errors.push(e)
    }
  }
  if (formError.errors.length > 0) {
    return Promise.reject(formError)
  }
  const method = isObject(params) && isString(params.method) ? params.method : ``
  if (method === Method.Update && keys(validatedObject).length === 0) {
    return Promise.reject({model: form.model, errors: [{
      model: `none`, type: ErrorType.emptyUpdate
    }]})
  }
  const requireOwner = getRequireOwner(form, method)
  if (isString(requireOwner)) {
    const idPath: any = params.idPath
    set(validatedObject, requireOwner, get(params.currentUser, idPath))
  }
  return Promise.resolve(validatedObject)
}

export default validateForm
