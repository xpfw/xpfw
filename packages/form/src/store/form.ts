import { get, set } from "lodash"
import { action, observable } from "mobx"
import { prependPrefix } from "../util/prefixMaker"

/**
 * Holds data relevant for Forms such as values and validation errors
 */
export class FormStoreClass {
  /**
   * MobX Observable containing all form data
   */
  @observable
  public formData: any = {}
  /**
   * MobX Observable containing all form errors
   */
  @observable
  public errors: any = {}
  /**
   * MobX Observable containing Loading states
   */
  @observable
  public loading: any = {}

  /**
   * Acquire a value from the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   * @param defaultValue value to return if it is not set in the form data
   */
  public getValue(valuePath?: string, prefix?: string, defaultValue?: any): any {
    return get(this.formData, prependPrefix(valuePath, prefix), defaultValue)
  }

  /**
   * Set a value in the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setValue(valuePath?: string, value?: any, prefix?: string) {
    set(this.formData, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getError(valuePath?: string, prefix?: string): any {
    return get(this.errors, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setError(valuePath?: string, value?: any, prefix?: string) {
    set(this.errors, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getLoading(valuePath?: string, prefix?: string): any {
    return get(this.loading, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setLoading(valuePath: string, value: boolean, prefix?: string) {
    set(this.loading, prependPrefix(valuePath, prefix), value)
  }
}

const FormStore = new FormStoreClass()
export default FormStore
