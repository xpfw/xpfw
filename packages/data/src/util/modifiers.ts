
import { executeForMethods, ModifyFunction } from "@xpfw/form"
import { get } from "lodash"
import dataOptions from "../options"
import UserStore from "../store/user"

const addUserId = (path: string, methods?: string[]) => {
  const userIdAdder: ModifyFunction = executeForMethods((value) => {
    value[path] = get(UserStore.getUser(), dataOptions.idPath)
    return Promise.resolve(value)
  }, methods)
  return userIdAdder
}
const addDefaultSort = (sortValue: any, methods?: string[]) => {
  const userIdAdder: ModifyFunction = executeForMethods((value) => {
    if (value.$sort == null) {
      value.$sort = sortValue
    }
    return Promise.resolve(value)
  }, methods)
  return userIdAdder
}

export {
  addUserId, addDefaultSort
}
