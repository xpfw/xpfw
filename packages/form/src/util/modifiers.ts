import { ModifyFunction } from "@xpfw/form"
import { get } from "lodash"
import { ExtendedJSONSchema } from "@xpfw/form"

const useModifier =
async (value: any, schema: ExtendedJSONSchema, method: string) => {
  let valToReturn = value
  let modifiers = schema.modify
  if (modifiers != null) {
    if (!Array.isArray(modifiers)) {
      modifiers = [modifiers]
    }
    for (const modifier of modifiers) {
      valToReturn = await modifier(valToReturn, schema, method)
    }
  }
  return valToReturn
}

const executeForMethods = (f: ModifyFunction, methods?: string[]) => {
  const modifier: ModifyFunction = (value, schema, method) => {
    if (methods == null || methods.indexOf(method) !== -1) {
      return f(value, schema, method)
    }
    return Promise.resolve(value)
  }
  return modifier
}

const addTimeStamp = (path: string, methods?: string[]) => {
  const timestampAdder: ModifyFunction = executeForMethods((value) => {
    value[path] = new Date()
    return Promise.resolve(value)
  }, methods)
  return timestampAdder
}

export {
  executeForMethods, addTimeStamp, useModifier
}
