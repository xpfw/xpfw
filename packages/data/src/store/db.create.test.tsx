import "isomorphic-fetch"
import * as MockDate from "mockdate"
MockDate.set(new Date(4, 2, 0))

import { FeathersClient } from "@xpfw/data-feathers"
import { toJS } from "@xpfw/data-tests"
import { addTimeStamp } from "@xpfw/form"
import { makeSubFields, NameField, NumberAndRequiredTextSchema, NumberField } from "@xpfw/form-tests"
import { getRandomApp } from "@xpfw/test-util"
import { get } from "lodash"
import BackendClient from "../client"
import { addUserId } from "../util/modifiers"
import DbStore from "./db"
import UserStore from "./user"

BackendClient.client = FeathersClient

test("DbStore Create Test", async () => {
  const s: any = NumberAndRequiredTextSchema.collection
  const appRef = await getRandomApp(s, false, BackendClient.client, true)
  const prefix = "createpref"
  const fields = makeSubFields(NumberAndRequiredTextSchema)
  fields[NameField.title].setValue("myText")
  fields[NumberField.title].setValue(420)
  expect(toJS(DbStore)).toMatchSnapshot("Before calling create")
  UserStore.user = {id: "myuserid"}
  const opts: any = {
    addCreatedAt: true,
    addBelongsTo: true
  }
  NumberAndRequiredTextSchema.modify = [addTimeStamp("createdAt"), addUserId("belongsTo")]
  const createdObject = await DbStore.create(NumberAndRequiredTextSchema)
  expect(toJS(DbStore)).toMatchSnapshot("After calling create")
  const col: any = NumberAndRequiredTextSchema.collection
  const getResult = await DbStore.getFromServer(get(createdObject, "id"), col)
  expect(toJS(getResult)).toMatchSnapshot("getresult")
  expect(getResult).toEqual(createdObject)
  expect(getResult).toEqual(DbStore.getCreateState(NumberAndRequiredTextSchema.title))

  NumberAndRequiredTextSchema.collection = "broken"
  const res = await DbStore.create(NumberAndRequiredTextSchema)
  expect(res).not.toBeNull()
  await appRef.cleanUp()
}, 10000)
