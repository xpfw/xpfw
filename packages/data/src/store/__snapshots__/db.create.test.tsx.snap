// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DbStore Create Test: After calling create 1`] = `
Object {
  "createState": Object {
    "formModel": Object {
      "belongsTo": "myuserid",
      "createdAt": "2011-05-05T14:20:43.337Z",
      "guestName": "myText",
      "id": 0,
      "myVal": 420,
    },
  },
  "currentlyEditing": "",
  "fetching": Object {},
  "getState": Object {},
  "lastFetch": Object {},
  "removeState": Object {},
  "updateState": Object {},
}
`;

exports[`DbStore Create Test: Before calling create 1`] = `
Object {
  "createState": Object {},
  "currentlyEditing": "",
  "fetching": Object {},
  "getState": Object {},
  "lastFetch": Object {},
  "removeState": Object {},
  "updateState": Object {},
}
`;

exports[`DbStore Create Test: getresult 1`] = `
Object {
  "belongsTo": "myuserid",
  "createdAt": "2011-05-05T14:20:43.337Z",
  "guestName": "myText",
  "id": 0,
  "myVal": 420,
}
`;
